### ISSUE TITLE
ipsec systemd-resolved support

### ISSUE THREAD

**Issue by Ypnose** [strongswan/strongswan#1354]
**System :**
 - OS: Ubuntu 22.04.1
 - Kernel version: e.g. 5.15
 - strongSwan version(s): 5.9.5
 - Tested/confirmed with the latest version: no

**Describe the bug**
In a IKEv2 roadwarrior setup, the client doesn't update DNS when the tunnel is launched with `ipsec start` and when systemd-resolved is used.

**To Reproduce**
1. Launch the tunnel with `ipsec start --no-fork`
2. The client is successfully connected
3. Internet access doesn't work, as the remote DNS server is not used

**Expected behavior**
The client should be able to resolve domains

**Logs/Backtraces**
`resolvectl status` still shows same DNS as if the client was not connected.

**Additional context**
macOS clients using the same server have no problems. `resolvctl` can be invoked as `resolvconf`, it doesn't seem to be called here. I have to use a wrapper like this :

	#!/bin/bash
	set -e

	trap 'clean_dns' ERR INT EXIT

	clean_dns() {
	    if [[ -e /etc/systemd/resolved.conf.d/ike.conf ]]; then
	        rm /etc/systemd/resolved.conf.d/ike.conf
	        systemctl restart systemd-resolved
	    fi
	}

	mkdir -p /etc/systemd/resolved.conf.d/
	printf '[Resolve]\nDNS=10.10.10.1\nDomains=my.domain.tld ~.\n' \
	    >/etc/systemd/resolved.conf.d/ike.conf
	( sleep 4 && systemctl restart systemd-resolved ) &
	sudo ipsec start --nofork

	exit

Enforcing DNS IP and domain is not convenient, especially when the server directly provides them.

**Comment by tobiasbrunner** [comment]
Is the _resolve_ plugin loaded? Is `resolveconf` available (and provided by `systemd`)? Are there any errors logged when DNS servers are handled? Although, since you are using Ubuntu 22.04, I don't think it's the same issue as #1353.

**Comment by Ypnose** [comment]
Thanks for your help, it is very much appreciated.
Yes, resolve plugin is loaded :

```
00[LIB] loaded plugins: charon test-vectors ldap pkcs11 aesni aes rc2 sha2 sha1 md5 mgf1 rdrand random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl gcrypt af-alg fips-prf gmp curve25519 agent chapoly xcbc cmac hmac ctr ccm gcm ntru drbg curl attr kernel-netlink resolve socket-default connmark forecast farp stroke updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters
```

 `resolvconf` is not directly available on the system as is, but as `resolvectl` is a multi-call [binary](https://www.freedesktop.org/software/systemd/man/resolvectl.html#Compatibility%20with%0A%20%20%20%20resolvconf(8)), creating a link called `resolvconf` to `resolvectl` allows us to see this message :

```
resolvconf -a INTERFACE < FILE
resolvconf -d INTERFACE

Register DNS server and domain configuration with systemd-resolved.

  -h --help     Show this help
     --version  Show package version
  -a            Register per-interface DNS server and domain data
  -d            Unregister per-interface DNS server and domain data
  -f            Ignore if specified interface does not exist
  -x            Send DNS traffic preferably over this interface

This is a compatibility alias for the resolvectl(1) tool, providing native
command line compatibility with the resolvconf(8) tool of various Linux
distributions and BSD systems. Some options supported by other implementations
are not supported and are ignored: -m, -p, -u. Various options supported by other
implementations are not supported and will cause the invocation to fail:
-I, -i, -l, -R, -r, -v, -V, --enable-updates, --disable-updates,
--updates-are-enabled.

See the resolvectl(1) man page for details.
```
Unfortunately, it doesn't seem to modify DNS accordingly, when I verify it with `resolvectl status`.

**Comment by tobiasbrunner** [comment]
Where did you create that symlink? Note that the _resolve_ plugin currently only checks in `/sbin`.

**Comment by Ypnose** [comment]
Good point. The link was in `/usr/bin/` instead of `/usr/sbin/`.
Now, `ipsec` sees `resolvconf` but I have this error :
```
resolvconf: Failed to set DNS configuration: Unit dbus-org.freedesktop.network1.service not found
```
and then the revert also fails.

This service is not present on the system in `/lib/systemd/system`, nor `/etc/systemd/system` and not found with `systemctl cat`. I wonder if it shouldn't be set through `dbus-org.freedesktop.resolve1.service`.

**Comment by Ypnose** [comment]
Yes, I think so. `systemd-resolved` has no dependency and only wants `nss-lookup.target`. Setting up DNS through `dbus-org.freedesktop.resolve1.service` as a fallback, sounds redudant to you ?

**Comment by tobiasbrunner** [comment]
You mean directly via D-Bus? I've currently no plans to implement that. And `resolvectl` is meant to do that already. So it might not  change anything if `systemd-resolved` is the one using `org.freedesktop.network1` (I think `resolvectl` only uses the `org.freedesktop.resolve1` interface).

**Comment by Ypnose** [comment]
Ok, I see. I was thinking I could write a `resolvconf` wrapper to dynamically handle DNS IP server and domains. I didn't check sources yet, but do you think something like this could be possible ? I just need to see how _resolve_ plugin passes values.

**Comment by tibz7** [comment]
There is a similar problem on fedora 38. 
If i start the vpn with ` strongswan start` the resolvectl is not updated. 
``` 
aron[94733]: 12[IKE] authentication of 'gateway.meteoblue.com' with EAP successful
août 03 15:34:10 T15p charon[94733]: 12[IKE] installing DNS server 172.20.101.9 via resolvconf
août 03 15:34:10 T15p charon[94733]: 12[IKE] resolvconf: Dropped protocol specifier '.ipsec' from 'lo.ipsec'. Using 'lo' (ifindex=1).
août 03 15:34:10 T15p charon[94733]: 12[IKE] resolvconf: Failed to set DNS configuration: Could not activate remote peer: activation request failed: unknown unit.
août 03 15:34:10 T15p charon[94733]: 12[IKE] adding DNS server failed
```
```
resolvectl dns                                                                                                                                                             
Global: 1.1.1.1 
Link 2 (enp0s31f6):
Link 3 (wlp0s20f3): 1.1.1.1 192.168.1.1 2a02:1210:8406:ab00:a2b5:49ff:fe87:67b0
Link 4 (docker0):
Link 20 (ipsec1):
Link 37 (ipsec0):
```

however If i start it with the UI ( network manager) it works 

```
resolvectl dns                                                                                                                                                             
Global: 1.1.1.1 
Link 2 (enp0s31f6):
Link 3 (wlp0s20f3): 1.1.1.1 192.168.1.1 2a02:1210:8406:ab00:a2b5:49ff:fe87:67b0
Link 4 (docker0):
Link 20 (ipsec1):
Link 37 (ipsec0): my.dns-server.com
```

Any idea why ? or how to solve that?

**Comment by jjakob** [comment]
While one can write a script that sets global DNS servers and search domains in systemd-resolved for tunnels from Strongswan, personally I have a different use case, I don't want the DNS servers from tunnels being applied globally (which would mean all DNS resolution would be forwarded to them). I want only the tunnel's specific search domains to be forwarded to the tunnel's DNS servers.

I think systemd-resolved doesn't support adding such configuration without a network interface to associate the configuration with. So if one uses tunnel-based connections, there is no separate network interface associated with them.

from the systemd-resolved manpage (they use the term "link" to refer to a network interface):
```
       •   If a name to look up matches (that is: is equal to or has as suffix) any of the configured routing domains (search
           or route-only) of any link, or the globally configured DNS settings, "best matching" routing domain is determined:
           the matching one with the most labels. The query is then sent to all DNS servers of any links or the globally
           configured DNS servers associated with this "best matching" routing domain. (Note that more than one link might have
           this same "best matching" routing domain configured, in which case the query is sent to all of them in parallel).
```
I tried using resolvectl to add DNS for a link name that doesn't exist on the system, but it wouldn't do it:
```
$ resolvectl dns ipsec_foo 192.0.2.3
Failed to resolve interface "ipsec_foo": No such device
```

If the connection is route-based and adds a network interface, it is possible to use `resolvectl`  to add search domains and DNS forwarders for it. So this is a limitation of systemd-resolved, it might be a good idea to check if this is the case with systemd-resolved developers and if it is, file a feature request for this functionality to be added to systemd-resolved.

In comparison, unbound can easily be configured to only forward specific zones to specific nameservers using stub zones, which can be done programatically with `unbound-control stub_add` or manually in the config file, for example:
```
server:
        domain-insecure: "example.com."
stub-zone:
        name: "example.com."
	stub-addr: 192.0.2.3
```
